docs/templates/tutorials/subclass-tuner.md

```diff
@@ -0,0 +1,132 @@
+
+### Example
+
+```python
+import kerastuner as kt
+import tensorflow as tf
+import tensorflow_datasets as tfds
+
+
+def build_model(hp):
+    """Builds a convolutional model."""
+    inputs = tf.keras.Input(shape=(28, 28, 1))
+    x = inputs
+    for i in range(hp.Int('conv_layers', 1, 3, default=3)):
+        x = tf.keras.layers.Conv2D(
+            filters=hp.Int('filters_' + str(i), 4, 32, step=4, default=8),
+            kernel_size=hp.Int('kernel_size_' + str(i), 3, 5),
+            activation='relu',
+            padding='same')(x)
+
+        if hp.Choice('pooling' + str(i), ['max', 'avg']) == 'max':
+            x = tf.keras.layers.MaxPooling2D()(x)
+        else:
+            x = tf.keras.layers.AveragePooling2D()(x)
+
+        x = tf.keras.layers.BatchNormalization()(x)
+        x = tf.keras.layers.ReLU()(x)
+
+    if hp.Choice('global_pooling', ['max', 'avg']) == 'max':
+        x = tf.keras.layers.GlobalMaxPooling2D()(x)
+    else:
+        x = tf.keras.layers.GlobalAveragePooling2D()(x)
+    outputs = tf.keras.layers.Dense(10, activation='softmax')(x)
+
+    model = tf.keras.Model(inputs, outputs)
+
+    optimizer = hp.Choice('optimizer', ['adam', 'sgd'])
+    model.compile(optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
+    return model
+
+
+class MyTuner(kt.Tuner):
+
+    def run_trial(self, trial, train_ds):
+        hp = trial.hyperparameters
+
+        # Hyperparameters can be added anywhere inside `run_trial`.
+        # When the first trial is run, they will take on their default values.
+        # Afterwards, they will be tuned by the `Oracle`.
+        train_ds = train_ds.batch(
+            hp.Int('batch_size', 32, 128, step=32, default=64))
+
+        model = self.hypermodel.build(trial.hyperparameters)
+        lr = hp.Float('learning_rate', 1e-4, 1e-2, sampling='log', default=1e-3)
+        optimizer = tf.keras.optimizers.Adam(lr)
+        epoch_loss_metric = tf.keras.metrics.Mean()
+
+        @tf.function
+        def run_train_step(data):
+            images = tf.dtypes.cast(data['image'], 'float32') / 255.
+            labels = data['label']
+            with tf.GradientTape() as tape:
+                logits = model(images)
+                loss = tf.keras.losses.sparse_categorical_crossentropy(
+                    labels, logits)
+                # Add any regularization losses.
+                if model.losses:
+                    loss += tf.math.add_n(model.losses)
+                gradients = tape.gradient(loss, model.trainable_variables)
+            optimizer.apply_gradients(zip(gradients, model.trainable_variables))
+            epoch_loss_metric.update_state(loss)
+            return loss
+
+        # `self.on_epoch_end` reports results to the `Oracle` and saves the
```

Comment:
It should be specified in the text above ("Typically, `Tuner.run_trial` is the only method that users need to override when subclassing `Tuner`.") that `run_trial` should call `self.on_epoch_end` and optionally other callback methods such as `on_epoch_begin`, etc. The list of available methods, and which ones are optional or not, should be made explicit.

Reply:
Done